src/psyclone/parse/module_info.py

```diff
@@ -0,0 +1,212 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2023, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+
+'''This module contains the ModuleInfo class, which is used to store
+and cache information about a module: the filename, source code (if requested)
+and the fparser tree (if requested).
+'''
+
+
+from fparser.common.readfortran import FortranStringReader
```

Comment:
These imports all add weight to the idea that perhaps this functionality belongs in fparser. We've been trying to isolate PSyclone's fparser dependence to within the PSyIR.

Reply:
See above for details. Otherwise: the main reason I use fparser here is:
1. it can handle constants_mod.F90 (with preprocessor directives), which psyir cannot do :(
2. it allows me to handle interfaces (which again PSyclone doesn't do as far as I can see) - and I need this for gungho.

So ideally I would like to get rid of the fparser layer entirely, but for now I need it.

For 1): in the LFRic environment filed are pre-processed when PSyclone sees them. Maybe we should add support for this - include both the .F90 (which would allow us to easily test different set of defines) and .f90 file, and some logic to prefer .f90 over .F90 files.

Reply:
As mentioned above, I've added the preprocessed files.